dev-tools/reconfigurator-cli/src/main.rs

```diff
@@ -529,8 +669,116 @@ fn cmd_blueprint_diff(
         .get(&blueprint2_id)
         .ok_or_else(|| anyhow!("no such blueprint: {}", blueprint2_id))?;
 
-    let diff = blueprint1.diff_sleds(&blueprint2);
-    Ok(Some(diff.display().to_string()))
+    let sled_diff = blueprint1.diff_sleds(&blueprint2).display().to_string();
+    swriteln!(rv, "{}", sled_diff);
+
+    // Diff'ing DNS is a little trickier.  First, compute what DNS should be for
+    // each blueprint.  To do that we need to construct a list of sleds suitable
+    // for the executor.
+    let sleds_by_id = make_sleds_by_id(&sim)?;
+    let internal_dns_config1 = blueprint_internal_dns_config(
+        &blueprint1,
+        &sleds_by_id,
+        &Default::default(),
+    )?;
+    let internal_dns_config2 = blueprint_internal_dns_config(
+        &blueprint2,
+        &sleds_by_id,
+        &Default::default(),
+    )?;
+    let dns_diff = DnsDiff::new(&internal_dns_config1, &internal_dns_config2)
+        .context("failed to assemble DNS diff")?;
+    swriteln!(rv, "internal DNS:\n{}", dns_diff);
+
+    let external_dns_config1 = blueprint_external_dns_config(
+        &blueprint1,
+        &sim.silo_names,
+        sim.external_dns_zone_name.clone(),
+    );
+    let external_dns_config2 = blueprint_external_dns_config(
+        &blueprint2,
+        &sim.silo_names,
+        sim.external_dns_zone_name.clone(),
+    );
+    let dns_diff = DnsDiff::new(&external_dns_config1, &external_dns_config2)
+        .context("failed to assemble external DNS diff")?;
+    swriteln!(rv, "external DNS:\n{}", dns_diff);
+
+    Ok(Some(rv))
+}
+
+fn make_sleds_by_id(
+    sim: &ReconfiguratorSim,
+) -> Result<BTreeMap<Uuid, nexus_reconfigurator_execution::Sled>, anyhow::Error>
+{
+    let collection = sim
+        .system
+        .to_collection_builder()
+        .context(
+            "unexpectedly failed to create collection for current set of sleds",
+        )?
+        .build();
+    let sleds_by_id: BTreeMap<_, _> = collection
+        .sled_agents
+        .iter()
+        .map(|(sled_id, sled_agent_info)| {
+            let sled = nexus_reconfigurator_execution::Sled::new(
```

Comment:
I was going to ask if this should consider the sled policy (e.g., to skip expunged sleds), but I don't think that's a property of collections, right? Any sled_agent present in a collection was not expunged at the time, by definition, right?

Reply:
This is a really good question.  There's definitely something a little fishy about this.   tl;dr: I think it doesn't affect what we're currently using this tool for, and fixing it is somewhat hard, so I think it's beyond the scope of this PR.

Background: we're constructing this list of sleds solely to be able to figure out what internal DNS records reconfigurator _would_ create for a given blueprint.  Reconfigurator only needs this list of sleds to figure out which sleds are scrimlets so that it can generate the switch zone services' DNS records.  Eventually it should generate sled agent DNS records too but we don't seem to do that today.

This code is being used in two contexts:

- when diff'ing two blueprints, to figure out how DNS would change between them
- when diff'ing a blueprint against an actual DNS config, to see how executing that blueprint would change DNS

But as you're observing, the actual DNS contents for a blueprint depends on the policy, not just the blueprint.  It's maybe misleading that reconfigurator-cli has an operation to diff blueprints without saying anything about the policy.  It should probably let you keep track of multiple policies and make you specify which policy you want.  You would _need_ this if you wanted to use these tools to see how DNS would change if you just expunged a sled.

I do think we should think about about this longer term.  For now I think the summary is:

- there is one state of the world that reconfigurator-cli knows about
- it's populated either with "sled-add" or "load" commands
- it is implicitly used wherever we need this information

This isn't the first place we implicitly use the policy like this.  We use it when generating an inventory and when generating a blueprint from the inventory, too.  But I think this is the first place where the user is referencing two points in time and might reasonably want two different policies.